Won't compile for wasm32-wasi

I'm trying to compile my canister to Wasi (I know this isn't an official target, but I am trying to get this to work and will be polyfilling the methods). Candid won't compile with the following errors. When I switch back to `wasm32-unknown-unknown` the compiler errors go away. From the errors it looks like this affects `0.8.4` and `0.9.0-beta.3`. The Cargo.lock files are identical when switching targets:

```
error[E0277]: the trait bound `BigUint: From<i64>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.9.0-beta.3/src/types/number.rs:192:42
    |
192 |                 Ok(Nat(BigUint::try_from(v).map_err(|_| {
    |                        ----------------- ^ the trait `From<i64>` is not implemented for `BigUint`
    |                        |
    |                        required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <BigUint as From<types::number::Nat>>
              <BigUint as From<u128>>
              <BigUint as From<u16>>
              <BigUint as From<u32>>
              <BigUint as From<u64>>
              <BigUint as From<u8>>
              <BigUint as From<usize>>
    = note: required for `i64` to implement `Into<BigUint>`
    = note: required for `BigUint` to implement `TryFrom<i64>`

error[E0277]: the trait bound `BigUint: From<i64>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.9.0-beta.3/src/types/number.rs:192:24
    |
192 |                 Ok(Nat(BigUint::try_from(v).map_err(|_| {
    |                        ^^^^^^^^^^^^^^^^^ the trait `From<i64>` is not implemented for `BigUint`
    |
    = help: the following other types implement trait `From<T>`:
              <BigUint as From<types::number::Nat>>
              <BigUint as From<u128>>
              <BigUint as From<u16>>
              <BigUint as From<u32>>
              <BigUint as From<u64>>
              <BigUint as From<u8>>
              <BigUint as From<usize>>
    = note: required for `i64` to implement `Into<BigUint>`
    = note: required for `BigUint` to implement `TryFrom<i64>`

error[E0277]: the trait bound `BigUint: From<i64>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/types/number.rs:198:42
    |
198 |                 Ok(Nat(BigUint::try_from(v).map_err(|_| {
    |                        ----------------- ^ the trait `From<i64>` is not implemented for `BigUint`
    |                        |
    |                        required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <BigUint as From<types::number::Nat>>
              <BigUint as From<u128>>
              <BigUint as From<u16>>
              <BigUint as From<u32>>
              <BigUint as From<u64>>
              <BigUint as From<u8>>
              <BigUint as From<usize>>
    = note: required for `i64` to implement `Into<BigUint>`
    = note: required for `BigUint` to implement `TryFrom<i64>`

error[E0277]: the trait bound `BigUint: From<i64>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/types/number.rs:198:24
    |
198 |                 Ok(Nat(BigUint::try_from(v).map_err(|_| {
    |                        ^^^^^^^^^^^^^^^^^ the trait `From<i64>` is not implemented for `BigUint`
    |
    = help: the following other types implement trait `From<T>`:
              <BigUint as From<types::number::Nat>>
              <BigUint as From<u128>>
              <BigUint as From<u16>>
              <BigUint as From<u32>>
              <BigUint as From<u64>>
              <BigUint as From<u8>>
              <BigUint as From<usize>>
    = note: required for `i64` to implement `Into<BigUint>`
    = note: required for `BigUint` to implement `TryFrom<i64>`

error[E0277]: the trait bound `i128: From<BigInt>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.9.0-beta.3/src/de.rs:504:23
    |
504 |                 int.0.try_into().map_err(Error::msg)?
    |                       ^^^^^^^^ the trait `From<BigInt>` is not implemented for `i128`
    |
    = help: the following other types implement trait `From<T>`:
              <i128 as From<NonZeroI128>>
              <i128 as From<bool>>
              <i128 as From<i16>>
              <i128 as From<i32>>
              <i128 as From<i64>>
              <i128 as From<i8>>
              <i128 as From<u16>>
              <i128 as From<u32>>
            and 2 others
    = note: required for `BigInt` to implement `Into<i128>`
    = note: required for `i128` to implement `TryFrom<BigInt>`
    = note: required for `BigInt` to implement `TryInto<i128>`

error[E0277]: the trait bound `i128: From<BigUint>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.9.0-beta.3/src/de.rs:508:23
    |
508 |                 nat.0.try_into().map_err(Error::msg)?
    |                       ^^^^^^^^ the trait `From<BigUint>` is not implemented for `i128`
    |
    = help: the following other types implement trait `From<T>`:
              <i128 as From<NonZeroI128>>
              <i128 as From<bool>>
              <i128 as From<i16>>
              <i128 as From<i32>>
              <i128 as From<i64>>
              <i128 as From<i8>>
              <i128 as From<u16>>
              <i128 as From<u32>>
            and 2 others
    = note: required for `BigUint` to implement `Into<i128>`
    = note: required for `i128` to implement `TryFrom<BigUint>`
    = note: required for `BigUint` to implement `TryInto<i128>`

error[E0277]: the trait bound `u128: From<BigUint>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.9.0-beta.3/src/de.rs:525:33
    |
525 |         let value: u128 = nat.0.try_into().map_err(Error::msg)?;
    |                                 ^^^^^^^^ the trait `From<BigUint>` is not implemented for `u128`
    |
    = help: the following other types implement trait `From<T>`:
              <u128 as From<Ipv6Addr>>
              <u128 as From<NonZeroU128>>
              <u128 as From<bool>>
              <u128 as From<char>>
              <u128 as From<u16>>
              <u128 as From<u32>>
              <u128 as From<u64>>
              <u128 as From<u8>>
    = note: required for `BigUint` to implement `Into<u128>`
    = note: required for `u128` to implement `TryFrom<BigUint>`
    = note: required for `BigUint` to implement `TryInto<u128>`

error[E0277]: the trait bound `i128: From<BigInt>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/de.rs:414:23
    |
414 |                 int.0.try_into().map_err(Error::msg)?
    |                       ^^^^^^^^ the trait `From<BigInt>` is not implemented for `i128`
    |
    = help: the following other types implement trait `From<T>`:
              <i128 as From<NonZeroI128>>
              <i128 as From<bool>>
              <i128 as From<i16>>
              <i128 as From<i32>>
              <i128 as From<i64>>
              <i128 as From<i8>>
              <i128 as From<u16>>
              <i128 as From<u32>>
            and 2 others
    = note: required for `BigInt` to implement `Into<i128>`
    = note: required for `i128` to implement `TryFrom<BigInt>`
    = note: required for `BigInt` to implement `TryInto<i128>`

error[E0277]: the trait bound `i128: From<BigUint>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/de.rs:418:23
    |
418 |                 nat.0.try_into().map_err(Error::msg)?
    |                       ^^^^^^^^ the trait `From<BigUint>` is not implemented for `i128`
    |
    = help: the following other types implement trait `From<T>`:
              <i128 as From<NonZeroI128>>
              <i128 as From<bool>>
              <i128 as From<i16>>
              <i128 as From<i32>>
              <i128 as From<i64>>
              <i128 as From<i8>>
              <i128 as From<u16>>
              <i128 as From<u32>>
            and 2 others
    = note: required for `BigUint` to implement `Into<i128>`
    = note: required for `i128` to implement `TryFrom<BigUint>`
    = note: required for `BigUint` to implement `TryInto<i128>`

error[E0277]: the trait bound `u128: From<BigUint>` is not satisfied
   --> /home/lastmjs/.config/kybra/1.68.2/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/de.rs:432:33
    |
432 |         let value: u128 = nat.0.try_into().map_err(Error::msg)?;
    |                                 ^^^^^^^^ the trait `From<BigUint>` is not implemented for `u128`
    |
    = help: the following other types implement trait `From<T>`:
              <u128 as From<Ipv6Addr>>
              <u128 as From<NonZeroU128>>
              <u128 as From<bool>>
              <u128 as From<char>>
              <u128 as From<u16>>
              <u128 as From<u32>>
              <u128 as From<u64>>
              <u128 as From<u8>>
    = note: required for `BigUint` to implement `Into<u128>`
    = note: required for `u128` to implement `TryFrom<BigUint>`
    = note: required for `BigUint` to implement `TryInto<u128>`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `candid` due to 5 previous errors
warning: build failed, waiting for other jobs to finish...
error: could not compile `candid` due to 5 previous errors
```

This very simple example will reproduce the issue:

`Cargo.toml`:

```toml
[package]
name = "test"
version = "0.0.0"
edition = "2018"

[dependencies]
candid="0.8.4"
```

`src/main.rs`:

```rust
fn main() {
    println!("{}", candid::Int::from(0));
}
```

```
rustup target add wasm32-wasi
cargo build --target wasm32-wasi
```

These errors are produced: 

```
   Compiling typenum v1.16.0
   Compiling memchr v2.5.0
   Compiling num-traits v0.2.15
   Compiling serde v1.0.162
   Compiling regex-syntax v0.7.1
   Compiling cfg-if v1.0.0
   Compiling unicode-width v0.1.10
   Compiling typed-arena v2.0.2
   Compiling lazy_static v1.4.0
   Compiling termcolor v1.2.0
   Compiling arrayvec v0.5.2
   Compiling thiserror v1.0.40
   Compiling crc32fast v1.3.2
   Compiling binread v2.2.0
   Compiling num_enum v0.5.11
   Compiling candid v0.8.4
   Compiling anyhow v1.0.71
   Compiling logos v0.12.1
   Compiling aho-corasick v1.0.1
   Compiling pretty v0.10.0
   Compiling codespan-reporting v0.11.1
   Compiling hex v0.4.3
   Compiling leb128 v0.2.5
   Compiling generic-array v0.14.7
   Compiling data-encoding v2.3.3
   Compiling num-integer v0.1.45
   Compiling byteorder v1.4.3
   Compiling block-buffer v0.10.4
   Compiling crypto-common v0.1.6
   Compiling digest v0.10.6
   Compiling sha2 v0.10.6
   Compiling regex v1.8.1
   Compiling lalrpop-util v0.19.12
   Compiling num-bigint v0.4.3
   Compiling serde_bytes v0.11.9
error[E0277]: the trait bound `BigUint: From<i64>` is not satisfied
   --> /home/lastmjs/.cargo/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/types/number.rs:198:42
    |
198 |                 Ok(Nat(BigUint::try_from(v).map_err(|_| {
    |                        ----------------- ^ the trait `From<i64>` is not implemented for `BigUint`
    |                        |
    |                        required by a bound introduced by this call
    |
    = help: the following other types implement trait `From<T>`:
              <BigUint as From<types::number::Nat>>
              <BigUint as From<u128>>
              <BigUint as From<u16>>
              <BigUint as From<u32>>
              <BigUint as From<u64>>
              <BigUint as From<u8>>
              <BigUint as From<usize>>
    = note: required for `i64` to implement `Into<BigUint>`
    = note: required for `BigUint` to implement `TryFrom<i64>`

error[E0277]: the trait bound `BigUint: From<i64>` is not satisfied
   --> /home/lastmjs/.cargo/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/types/number.rs:198:24
    |
198 |                 Ok(Nat(BigUint::try_from(v).map_err(|_| {
    |                        ^^^^^^^^^^^^^^^^^ the trait `From<i64>` is not implemented for `BigUint`
    |
    = help: the following other types implement trait `From<T>`:
              <BigUint as From<types::number::Nat>>
              <BigUint as From<u128>>
              <BigUint as From<u16>>
              <BigUint as From<u32>>
              <BigUint as From<u64>>
              <BigUint as From<u8>>
              <BigUint as From<usize>>
    = note: required for `i64` to implement `Into<BigUint>`
    = note: required for `BigUint` to implement `TryFrom<i64>`

error[E0277]: the trait bound `i128: From<BigInt>` is not satisfied
   --> /home/lastmjs/.cargo/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/de.rs:414:23
    |
414 |                 int.0.try_into().map_err(Error::msg)?
    |                       ^^^^^^^^ the trait `From<BigInt>` is not implemented for `i128`
    |
    = help: the following other types implement trait `From<T>`:
              <i128 as From<NonZeroI128>>
              <i128 as From<bool>>
              <i128 as From<i16>>
              <i128 as From<i32>>
              <i128 as From<i64>>
              <i128 as From<i8>>
              <i128 as From<u16>>
              <i128 as From<u32>>
            and 2 others
    = note: required for `BigInt` to implement `Into<i128>`
    = note: required for `i128` to implement `TryFrom<BigInt>`
    = note: required for `BigInt` to implement `TryInto<i128>`

error[E0277]: the trait bound `i128: From<BigUint>` is not satisfied
   --> /home/lastmjs/.cargo/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/de.rs:418:23
    |
418 |                 nat.0.try_into().map_err(Error::msg)?
    |                       ^^^^^^^^ the trait `From<BigUint>` is not implemented for `i128`
    |
    = help: the following other types implement trait `From<T>`:
              <i128 as From<NonZeroI128>>
              <i128 as From<bool>>
              <i128 as From<i16>>
              <i128 as From<i32>>
              <i128 as From<i64>>
              <i128 as From<i8>>
              <i128 as From<u16>>
              <i128 as From<u32>>
            and 2 others
    = note: required for `BigUint` to implement `Into<i128>`
    = note: required for `i128` to implement `TryFrom<BigUint>`
    = note: required for `BigUint` to implement `TryInto<i128>`

error[E0277]: the trait bound `u128: From<BigUint>` is not satisfied
   --> /home/lastmjs/.cargo/registry/src/github.com-1ecc6299db9ec823/candid-0.8.4/src/de.rs:432:33
    |
432 |         let value: u128 = nat.0.try_into().map_err(Error::msg)?;
    |                                 ^^^^^^^^ the trait `From<BigUint>` is not implemented for `u128`
    |
    = help: the following other types implement trait `From<T>`:
              <u128 as From<Ipv6Addr>>
              <u128 as From<NonZeroU128>>
              <u128 as From<bool>>
              <u128 as From<char>>
              <u128 as From<u16>>
              <u128 as From<u32>>
              <u128 as From<u64>>
              <u128 as From<u8>>
    = note: required for `BigUint` to implement `Into<u128>`
    = note: required for `u128` to implement `TryFrom<BigUint>`
    = note: required for `BigUint` to implement `TryInto<u128>`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `candid` due to 5 previous errors
```

--- a/rust/candid/Cargo.toml
+++ b/rust/candid/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "candid"
-version = "0.9.2"
+version = "0.9.3"
 edition = "2021"
 authors = ["DFINITY Team"]
 description = "Candid is an interface description language (IDL) for interacting with canisters running on the Internet Computer."

--- a/rust/candid/src/de.rs
+++ b/rust/candid/src/de.rs
@@ -505,16 +505,21 @@ impl<'de, 'a> de::Deserializer<'de> for &'a mut Deserializer<'de> {
     where
         V: Visitor<'de>,
     {
+        use num_traits::ToPrimitive;
         self.unroll_type()?;
         assert!(*self.expect_type == TypeInner::Int);
         let value: i128 = match self.wire_type.as_ref() {
             TypeInner::Int => {
                 let int = Int::decode(&mut self.input).map_err(Error::msg)?;
-                int.0.try_into().map_err(Error::msg)?
+                int.0
+                    .to_i128()
+                    .ok_or_else(|| Error::msg("Cannot convert int to i128"))?
             }
             TypeInner::Nat => {
                 let nat = Nat::decode(&mut self.input).map_err(Error::msg)?;
-                nat.0.try_into().map_err(Error::msg)?
+                nat.0
+                    .to_i128()
+                    .ok_or_else(|| Error::msg("Cannot convert nat to i128"))?
             }
             t => return Err(Error::subtype(format!("{t} cannot be deserialized to int"))),
         };
@@ -524,13 +529,17 @@ impl<'de, 'a> de::Deserializer<'de> for &'a mut Deserializer<'de> {
     where
         V: Visitor<'de>,
     {
+        use num_traits::ToPrimitive;
         self.unroll_type()?;
         check!(
             *self.expect_type == TypeInner::Nat && *self.wire_type == TypeInner::Nat,
             "nat"
         );
         let nat = Nat::decode(&mut self.input).map_err(Error::msg)?;
-        let value: u128 = nat.0.try_into().map_err(Error::msg)?;
+        let value = nat
+            .0
+            .to_u128()
+            .ok_or_else(|| Error::msg("Cannot convert nat to u128"))?;
         visitor.visit_u128(value)
     }
     fn deserialize_unit<V>(self, visitor: V) -> Result<V::Value>

--- a/rust/candid/src/types/number.rs
+++ b/rust/candid/src/types/number.rs
@@ -188,9 +188,10 @@ impl<'de> Deserialize<'de> for Nat {
                 formatter.write_str("Nat value")
             }
             fn visit_i64<E: de::Error>(self, v: i64) -> Result<Nat, E> {
-                Ok(Nat(BigUint::try_from(v).map_err(|_| {
-                    de::Error::custom("int cannot be converted to nat")
-                })?))
+                use num_bigint::ToBigUint;
+                v.to_biguint()
+                    .ok_or_else(|| de::Error::custom("i64 cannot be converted to nat"))
+                    .map(Nat)
             }
             fn visit_u64<E>(self, v: u64) -> Result<Nat, E> {
                 Ok(Nat::from(v))

